homeassistant/components/mqtt/mixins.py

```diff
@@ -823,105 +823,99 @@ async def async_added_to_hass(self) -> None:
+
+    async def _async_process_discovery_update(
+        self,
+        payload: MQTTDiscoveryPayload,
+        discovery_update: Callable[[MQTTDiscoveryPayload], Coroutine[Any, Any, None]],
+        discovery_data: DiscoveryInfoType,
+    ) -> None:
+        """Process discovery update."""
+        try:
+            await discovery_update(payload)
+        finally:
             send_discovery_done(self.hass, discovery_data)
 
-        @callback
-        def discovery_callback(payload: MQTTDiscoveryPayload) -> None:
-            """Handle discovery update.
+    async def _async_process_discovery_update_and_remove(self) -> None:
+        """Process discovery update and remove entity."""
+        if TYPE_CHECKING:
+            assert self._discovery_data
+        self._cleanup_discovery_on_remove()
+        await self._async_remove_state_and_registry_entry()
+        send_discovery_done(self.hass, self._discovery_data)
 
-            If the payload has changed we will create a task to
-            do the discovery update.
+    @callback
+    def _async_discovery_callback(self, payload: MQTTDiscoveryPayload) -> None:
+        """Handle discovery update.
 
-            As this callback can fire when nothing has changed, this
-            is a normal function to avoid task creation until it is needed.
-            """
-            _LOGGER.debug(
-                "Got update for entity with hash: %s '%s'",
-                discovery_hash,
-                payload,
+        If the payload has changed we will create a task to
+        do the discovery update.
+
+        As this callback can fire when nothing has changed, this
+        is a normal function to avoid task creation until it is needed.
+        """
+        if TYPE_CHECKING:
+            assert self._discovery_data
+        discovery_hash: tuple[str, str] = self._discovery_data[ATTR_DISCOVERY_HASH]
+        _LOGGER.debug(
+            "Got update for entity with hash: %s '%s'",
+            discovery_hash,
+            payload,
+        )
+        old_payload: DiscoveryInfoType
+        old_payload = self._discovery_data[ATTR_DISCOVERY_PAYLOAD]
+        debug_info.update_entity_discovery_data(self.hass, payload, self.entity_id)
+        if not payload:
+            # Empty payload: Remove component
+            _LOGGER.info("Removing component: %s", self.entity_id)
+            self.hass.async_create_task(
+                self._async_process_discovery_update_and_remove()
             )
-            if TYPE_CHECKING:
-                assert self._discovery_data
-            old_payload: DiscoveryInfoType
-            old_payload = self._discovery_data[ATTR_DISCOVERY_PAYLOAD]
-            debug_info.update_entity_discovery_data(self.hass, payload, self.entity_id)
-            if not payload:
-                # Empty payload: Remove component
-                _LOGGER.info("Removing component: %s", self.entity_id)
+        elif self._discovery_update:
+            if old_payload != payload:
```

Comment:
This replied on the side effect of `debug_info.update_entity_discovery_data` updating the payload.  I changed it to compare the old and new payload instead as the code was confusing to me